// models/Mail/index.js
"use strict";

const nodemailer = require('nodemailer');

const { DataModel } = require('lib/decorators');
const { customErr } = require('lib/error');
const { 
  safeEmail,
  safeNum,
  safeStr,
} = require('lib/safe');

const Mail = function() {
/**
 * Decorates this function
 */
  DataModel.call(this);
/**
 * Name of collection
 * to be stored in DB
 */
  this.collection = 'mail';
/**
 * A unique property to use
 * when checking for duplicates
 */
  this.uniqueVal = ''; // TODO
/**
 * Properties unique to this model
 */
  this.host = {
    value   : '',
    required: true,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.username = {
    value   : '',
    required: true,
    safe    : function(email) {
      return safeEmail(email);
    }
  }
  this.password = {
    value   : '',
    required: true,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.port = {
    value   : '',
    required: true,
    safe    : function(num) {
      return safeNum(num);
    }
  }
  this.secure = {
    value   : '',
    required: false,
    safe    : function(bool) {
      return safeBool(bool);
    }
  }
  this.requireTLS = {
    value   : '',
    required: false,
    safe    : function(bool) {
      return safeBool(bool);
    }
  }
  this.from = {
    value   : '',
    required: true,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.to = {
    value   : '',
    required: true,
    safe    : function(email) {
      return safeEmail(email);
    }
  }
  this.subject = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.text = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.html = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.send = function(callback) {
    const dataObj = this.getObject();

    Object.keys(dataObj).forEach( function(prop) {
      if (prop.required === true)
        if (prop.value === '') {
          this.reset();
          callback(customErr('Missing Required Value'))
        }
    }.bind(this));

    nodemailer.createTransport({
      host      : this.host,
      secure    : this.secure,
      requireTLS: this.requireTLS,
      port      : this.port,
      auth      : {
        user: this.username,
        pass: this.password
      }
    }.bind(this)).sendMail({
      from   : this.from,
      to     : this.to,
      subject: this.subject,
      text   : this.text,
      html   : this.html,
    }.bind(this), (error, info) => {
      if (error)
        return callback(error)
      return callback(null, info);
    });
  }
}

module.exports = Mail;
