// models/Property/index.js
"use strict";

const {
  DataModel,
  Address,
} = require('lib/decorators');
const { 
  safeNum,
  safeBool,
  safeStr,
  safeYear,
} = require('lib/safe');

const Property = function() {
/**
 * Decorates this function
 */
  DataModel.call(this);
/**
 * Inherit properties street, city,
 * state, and zip from Address.
 */
  Address.call(this);
/**
 * Name of collection
 * to be stored in DB
 */
  this.collection = 'property';
/**
 * A unique property to use
 * when checking for duplicates
 */
  this.uniqueVal = 'street';
/**
 * Properties unique to this model
 */
  this.mainImage = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.type = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.street = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }        
  }
  this.city = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.state = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.zip = {
    value   : '',
    required: false,
    safe    : function(num) {
      return safeNum(num);
    }
  }
  this.stories = {
    value   : '',
    required: false,
    safe    : function(num) {
      return safeNum(num);
    }
  }
  this.rent = {
    value   : '',
    required: false,
    safe    : function(num) {
      return safeNum(num);
    }
  }
  this.occupied = {
    value   : '',
    required: false,
    safe    : function(bool) {
      return safeBool(bool);
    }        
  }
  this.occupants = {
    value   : '',
    required: false,
    safe    : function(str) {
      return safeStr(str);
    }
  }
  this.sqft = {
    value   : '',
    required: false,
    safe    : function(num) {
      return safeNum(num);
    }
  }
  this.year = {
    value   : '',
    required: false,
    safe    : function(num) {
      return safeYear(num);
    }
  }
  this.washer = {
    value   : '',
    required: false,
    safe    : function(bool) {
      return safeBool(bool);
    }        
  }
  this.dryer = {
    value   : '',
    required: false,
    safe    : function(bool) {
      return safeBool(bool);
    }        
  }
  this.garage = {
    value   : '',
    required: false,
    safe    : function(bool) {
      return safeBool(bool);
    }        
  }
  this.basement = {
    value   : '',
    required: false,
    safe    : function(bool) {
      return safeBool(bool);
    }        
  }
  this.fence = {
    value   : '',
    required: false,
    safe    : function(bool) {
      return safeBool(bool);
    }        
  }
}

module.exports = Property;
